editor/src/components/editor/store/collaborative-editing.ts

```diff
@@ -0,0 +1,359 @@
+import type {
+  CollabTextFile,
+  CollabTextFileTopLevelElements,
+  CollaborativeEditingSupportSession,
+} from './editor-state'
+import { type CollaborativeEditingSupport } from './editor-state'
+import {
+  writeProjectFileChange,
+  type ProjectChanges,
+  type ProjectFileChange,
+  deletePathChange,
+  ensureDirectoryExistsChange,
+} from './vscode-changes'
+import type { ProjectContentsTree } from '../../../components/assets'
+import {
+  walkContentsTree,
+  type ProjectContentTreeRoot,
+  getProjectFileFromTree,
+  isProjectContentFile,
+  zipContentsTree,
+  addFileToProjectContents,
+} from '../../../components/assets'
+import type { EditorAction, EditorDispatch } from '../action-types'
+import {
+  deleteFile,
+  updateProjectContents,
+  updateTopLevelElements,
+} from '../actions/action-creators'
+import { assertNever, isBrowserEnvironment } from '../../../core/shared/utils'
+import type { ParseSuccess } from '../../../core/shared/project-file-types'
+import {
+  RevisionsState,
+  isTextFile,
+  parseSuccess,
+  textFile,
+  textFileContents,
+} from '../../../core/shared/project-file-types'
+import {
+  ParsedTextFileKeepDeepEquality,
+  TopLevelElementKeepDeepEquality,
+} from './store-deep-equality-instances'
+import * as Y from 'yjs'
+import type { TopLevelElement } from '../../../core/shared/element-template'
+import { isFeatureEnabled } from '../../../utils/feature-switches'
+
+export function collateCollaborativeProjectChanges(
+  oldContents: ProjectContentTreeRoot,
+  newContents: ProjectContentTreeRoot,
+): Array<ProjectFileChange> {
+  let changesToProcess: Array<ProjectFileChange> = []
+
+  function applyChanges(
+    fullPath: string,
+    firstContents: ProjectContentsTree,
+    secondContents: ProjectContentsTree,
+  ): boolean {
+    if (isProjectContentFile(firstContents)) {
```

Comment:
sad emoji that this is not happy-path-to-the-left

Reply:
There's not really a single happy path in this however.